### ISSUE TITLE
Issue running AVS Sample app on Samsung Android phone

### ISSUE THREAD

**Issue by smirkhani** [alexa/avs-device-sdk#1834]
**IMPORTANT**: Before you create an issue, please take a look at our [Issue Reporting Guide](https://github.com/alexa/avs-device-sdk/wiki/Issue-Reporting-Guide).

## Briefly summarize your issue:

I followed the instructions per this link: https://developer.amazon.com/en-US/docs/alexa/avs-device-sdk/android.html
But cannot run the sample app and get some runtime errors.
I should mention that my phone is not rooted and consequently I did not run the 'adb root'.

I read somewhere that we should not set the LD_LIBRARY_PATH environment variable anymore. Since the issue is associated with finding some symbols in .so files I though that might be the issue and I added the android's shared library path to the LD_LIBRARY_PATH but the issue is still there.  

## What is the expected behavior?

I was expecting that the Sample app actually runs

## What behavior are you observing?

cannot locate symbol "SSL_ctrl" referenced by "/data/local/tmp/myproject/lib/libavformat.so"...

## Provide the steps to reproduce the issue, if applicable:

Exactly the same steps proved in the link. 

### What version of the AVS Device SDK are you using?
      <x.y.z>
AVS SDK version is 1.21
my dev machine is Ubuntu 18
my phone is a Samsung not 9 with Android v.11

### Tell us what hardware you're using:
- [x ] Desktop / Laptop
- [ ] Raspberry Pi
- [ ] Other - tell us more:

### Tell us about your OS (Type & version):
- [x ] Linux
- [ ] MacOS 
- [ ] Raspbian Stretch
- [ ] Raspbian Jessy
- [ ] Other - tell us more:

**Comment by caleighatamazon** [comment]
Hi @smirkhani , I have a few initial thoughts of what might be going wrong.

1. Did the Android setup script run without error?
2. Did you update `androidConfig.txt` so that the target system and API level match that of your phone? It's possible the shared libraries were not compiled correctly for your phone.
3. Can you list the .so files found in `/data/local/tmp/myproject/lib` on your phone? Let's check that all the expected .so files are there.

**Comment by smirkhani** [comment]
Tnx for the prompt reply

>1-Did the Android setup script run without error?
Not really - I had to do the following three changes in my environment to make it work

a-CPP (preprocessor) in toolchain was missing in my environment - I removed the following line in the android.sh to avoid this issue

#export CPP="${TOOLCHAIN}/bin/${TOOLCHAIN_HOST}-cpp"

b- some basic math functions were not working because of the include dependencies-
I did some research and ended up removing the following line from avs-device-sdk/ApplicationUtilities/AndroidUtilities/src/CMakeLists.txt

"${ANDROID_NDK}/sysroot/usr/include"

c- for some reason the libc++ for arm64 was not generated in the directory that the scripts expected but I found it somewhere else and linked it

>2-Did you update androidConfig.txt so that the target system and API level match that of your phone? It's possible the shared libraries were not compiled correctly for your phone.

Yes - I updated the androidConfig.txt in the project root directory.
Here is the content of this file:
# Fill the lines below to configure android build.
PLATFORM="android"
API_LEVEL="30"
TARGET_SYSTEM="arm64"
DEVICE_INSTALL_PATH="/data/local/tmp/myproject"
BUILD_TYPE="debug"

BTW - I just noticed that there is another androidConfig.txt under ./avs-device-sdk/tools/Install/
I did not update that. Should I?

>3-Can you list the .so files found in /data/local/tmp/myproject/lib on your phone? Let's check that all the expected .so files are there.

Here is the list:
libACL.so                    libDeviceSettings.so            libacsdkAndroidApplicationAudioPipelineFactory.so libacsdkShared.so          
libADSL.so                   libEndpoints.so                 libacsdkAudioPlayer.so                            libacsdkShutdownManager.so 
libAFML.so                   libInteractionModel.so          libacsdkBluetooth.so                              libacsdkStartupManager.so  
libAIP.so                    libInterruptModel.so            libacsdkCBLAuthorizationDelegate.so               libavcodec.so              
libAVSCommon.so              libKWD.so                       libacsdkCore.so                                   libavdevice.so             
libAVSGatewayManager.so      libLibPreviewAlexaClient.so     libacsdkDefaultInternetConnectionMonitor.so       libavfilter.so             
libAVSSystem.so              libLibSampleApp.so              libacsdkDefaultSampleApplicationOptions.so        libavformat.so             
libAlexa.so                  libPlaybackController.so        libacsdkDoNotDisturb.so                           libavutil.so               
libAndroidSLESMediaPlayer.so libPlaylistParser.so            libacsdkEqualizer.so                              libc++_shared.so           
libAndroidUtilities.so       libRegistrationManager.so       libacsdkEqualizerImplementations.so               libcrypto.a                
libApiGateway.so             libSDKComponent.so              libacsdkExternalMediaPlayer.so                    libcurl.la                 
libAudioResources.so         libSQLiteStorage.so             libacsdkExternalMediaPlayerInterfaces.so          libnghttp2.a               
libCBLAuthDelegate.so        libSoftwareComponentReporter.so libacsdkLibcurlAlexaCommunications.so             libnghttp2.la              
libCapabilitiesDelegate.so   libSpeakerManager.so            libacsdkLibcurlHTTPContentFetcher.so              libsqlite3.la              
libCaptions.so               libSpeechEncoder.so             libacsdkManufactory.so                            libsqlite3.so              
libCaptionsComponent.so      libSpeechSynthesizer.so         libacsdkMultiRoomMusic.so                         libssl.a                   
libCaptionsLib.so            libSynchronizeStateSender.so    libacsdkNotifications.so                          libswresample.so           
libCertifiedSender.so        libSystemSoundPlayer.so         libacsdkNullMetricRecorder.so                     libswscale.so              
libContextManager.so         libTemplateRuntime.so           libacsdkPostConnectOperationProviderRegistrar.so  libz.so                    
libDefaultClient.so          libacsdkAlerts.so               libacsdkSampleApplicationCBLAuthRequester.so      pkgconfig

**Comment by caleighatamazon** [comment]
Hi @smirkhani , thanks for this information. This is definitely odd! The modifications you had to make suggest something has gone wrong earlier in your install process.

Did you allow the script to install the Android NDK for you, or did you install and use your own NDK? In the `android.sh` script, that variable is called `NDK_PACKAGE`.

The `android.sh` script actually only downloads either NDK 15 or 16 (so quite dated versions of the NDK). In fact, NDKr16 only supports up to API 27, not 30. You should not have been able to target `API_LEVEL=30` in your `androidConfig.txt` without encountering errors, if you had been using the NDK version our script downloads. (Remember that `API_LEVEL` just indicates the *minimum* API that you desire to support). You'd probably see an error like this (at least, this is the error I get when I try it!):

```
$HOME/Android/ndk/ndk-bundle/android-ndk-r16/platforms/android-30. Valid platforms:
android-16
android-27
android-15
android-22
android-26
android-23
android-18
NOTICE
android-19
android-24
android-14
android-17
android-21
repo.prop
```

If you're using a more recent version of the Android NDK, you may very well have run into the errors you report in your issue. I think you may have to dig a bit deeper to resolve those, though; for example, I don't think it is a good idea to comment out the `CPP` export. Your version of the NDK will have that somewhere, definitely, even if it is in a different location than our script expects. Unfortunately our `android.sh` helper script does not support more recent versions of the NDK.

I notice that you're missing `libcrypto.so` in the list of shared libraries you pasted above, which could explain the original runtime error you saw.

Anyway, let me know what you think of the above. If you are just using the Android NDK r16 that is downloaded by the script, then something else must be going wrong.

**Comment by smirkhani** [comment]
Ok - tnx for the clarification.

I created a new repo and changed my API level to 27 and then I relaunched the build. The whole build and installation gone through without any issue.

But, when I try to run the startsample.sh I get the following error:

```WARNING: linker: Warning: "/data/local/tmp/myproject/lib/libavutil.so" unused DT entry: DT_RPATH (type 0xf arg 0x2a48) (ignoring)
CANNOT LINK EXECUTABLE "/data/local/tmp/myproject/bin/SampleApp": cannot locate symbol "EVP_MD_CTX_init" referenced by "/system/lib64/libsqlite.so"...```

Please note that the above WARNING repeats for many other .so files but I only listed the last one here

**Comment by caleighatamazon** [comment]
Hi @smirkhani , how are you trying to run the Sample App, by the way?

I will see if I can reproduce this problem locally in the meantime.

**Comment by smirkhani** [comment]
I just ran source 

```
startsample.sh 
```
Appreciate your help. 

I am not an expert on Android but maybe my phone Android version or the API level is not compatible with your package.

**Comment by smirkhani** [comment]
Hi @caleighatamazon 
I still could not solve this problem yet.

I have another question please. Is AVS sample app even targeted for Android phones (the recent ones)? If yes, I was wondering which one of the phones (including Android version / API version) you guys think is compatible with AVS?  Tnx alot

**Comment by mvelegon-amzn** [comment]
Hi @smirkhani sorry for the delayed response. I am still investigating the issue at the moment. Will update you once I find something.

**Comment by mvelegon-amzn** [comment]
It is possible the android.sh script we run is not correctly cross compiling the dependencies for your Android phone's platform.

Can you post a log that shows the full output of android.sh to see if it can give us some clues. I also found some information here on this SO post: https://stackoverflow.com/questions/33206409/unused-dt-entry-type-0x1d-arg

**Comment by mvelegon-amzn** [comment]
Quick update. We were able to reproduce this issue locally will be posting an update once we have more information.

**Comment by mvelegon-amzn** [comment]
Hi @smirkhani Can you run this command before starting the SampleApp? Please let us know if it fixes the issue

`unset LD_LIBRARY_PATH`

**Comment by smirkhani** [comment]
Hi @caleighatamazon

Here are two files that contain the build log.
[avsAndroidSampleAppBuildLog1.log](https://github.com/alexa/avs-device-sdk/files/5616397/avsAndroidSampleAppBuildLog1.log)
[avsAndroidSampleAppBuildLog2.log](https://github.com/alexa/avs-device-sdk/files/5616404/avsAndroidSampleAppBuildLog2.log)

The only reason I have two log files is because I forgot to connect my phone in the first run and as you can see I had to stop the build and rerun build again. 

After the build I changed the startsample.sh to the following 
```
#  adb shell LD_LIBRARY_PATH=/data/local/tmp/lexieAVS/lib       /data/local/tmp/lexieAVS/bin/SampleApp       /data/local/tmp/lexieAVS/AlexaCl
ientSDKConfig.json DEBUG9
  adb shell   /data/local/tmp/lexieAVS/bin/SampleApp       /data/local/tmp/lexieAVS/AlexaClientSDKConfig.json DEBUG9
```

But still get the following error when I run the startsample.sh:

```
WARNING: linker: Warning: "/data/local/tmp/lexieAVS/bin/SampleApp" unused DT entry: DT_RPATH (type 0xf arg 0x39a7) (ignoring)
CANNOT LINK EXECUTABLE "/data/local/tmp/lexieAVS/bin/SampleApp": library "libLibSampleApp.so" not found
```

**Comment by smirkhani** [comment]
Hi @caleighatamazon - I was wondering if there is any update on this. Tnx

**Comment by smirkhani** [comment]
I used another phone - Pixel 3 with Android 11 
This time I actually run the sampleapp after building 
But when I enter 't' and ask for 'what time is it?', the app keeps telling me 'invalid option'.

**Comment by valeri-atamaniouk** [comment]
@smirkhani Hello, does the issue still exist? Can you try with the latest SDK release?

**Comment by kclchan** [comment]
I am closing this issue due to inactivity.  If you have further questions, please open a new ticket and refer to this closed one.
